File: java/design_pattern/observer/MainObserver.java
/**
 * Comprehensive demonstration of the Observer Pattern in Java
 *
 * The Observer pattern defines a one-to-many dependency between objects so that
 * when one object changes state, all its dependents are notified and updated
 * automatically.
 */

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// Observer interface
interface Observer {
    void update(String event, Object data);
}

// Subject interface
interface Subject {
    void attach(Observer observer);
    void detach(Observer observer);
    void notifyObservers(String event, Object data);
}

// Concrete Subject - News Agency
class NewsAgency implements Subject {
    private List<Observer> observers = new ArrayList<>();
    private String latestNews;

    @Override
    public void attach(Observer observer) {
        observers.add(observer);
    }

    @Override
    public void detach(Observer observer) {
        observers.remove(observer);
    }

    @Override
    public void notifyObservers(String event, Object data) {
        for (Observer observer : observers) {
            observer.update(event, data);
        }
    }

    public void publishNews(String news) {
        this.latestNews = news;
        System.out.println("  [NewsAgency] Publishing: " + news);
        notifyObservers("NEWS", news);
    }
}

// Concrete Observers
class NewsChannel implements Observer {
    private String name;

    public NewsChannel(String name) {
        this.name = name;
    }

    @Override
    public void update(String event, Object data) {
        System.out.println("  [" + name + "] Breaking news: " + data);
    }
}

class NewsWebsite implements Observer {
    private String name;

    public NewsWebsite(String name) {
        this.name = name;
    }

    @Override
    public void update(String event, Object data) {
        System.out.println("  [" + name + "] Website updated with: " + data);
    }
}

class MobileApp implements Observer {
    private String name;

    public MobileApp(String name) {
        this.name = name;
    }

    @Override
    public void update(String event, Object data) {
        System.out.println("  [" + name + "] Push notification: " + data);
    }
}

// Event-based Observer with multiple event types

interface EventListener {
    void onEvent(String eventType, Map<String, Object> data);
}

class EventManager {
    private Map<String, List<EventListener>> listeners = new HashMap<>();

    public EventManager(String... eventTypes) {
        for (String type : eventTypes) {
            listeners.put(type, new ArrayList<>());
        }
    }

    public void subscribe(String eventType, EventListener listener) {
        listeners.get(eventType).add(listener);
    }

    public void unsubscribe(String eventType, EventListener listener) {
        listeners.get(eventType).remove(listener);
    }

    public void notify(String eventType, Map<String, Object> data) {
        for (EventListener listener : listeners.get(eventType)) {
            listener.onEvent(eventType, data);
        }
    }
}

class FileEditor {
    private EventManager events;
    private String filename;

    public FileEditor() {
        events = new EventManager("open", "save", "close");
    }

    public EventManager getEvents() {
        return events;
    }

    public void openFile(String filename) {
        this.filename = filename;
        Map<String, Object> data = new HashMap<>();
        data.put("filename", filename);
        System.out.println("  [FileEditor] Opened: " + filename);
        events.notify("open", data);
    }

    public void saveFile() {
        Map<String, Object> data = new HashMap<>();
        data.put("filename", filename);
        System.out.println("  [FileEditor] Saved: " + filename);
        events.notify("save", data);
    }

    public void closeFile() {
        Map<String, Object> data = new HashMap<>();
        data.put("filename", filename);
        System.out.println("  [FileEditor] Closed: " + filename);
        events.notify("close", data);
        filename = null;
    }
}

class LoggingListener implements EventListener {
    private String logFile;

    public LoggingListener(String logFile) {
        this.logFile = logFile;
    }

    @Override
    public void onEvent(String eventType, Map<String, Object> data) {
        System.out.println("  [Logger] Writing to " + logFile + ": " +
                          eventType + " - " + data.get("filename"));
    }
}

class EmailListener implements EventListener {
    private String email;

    public EmailListener(String email) {
        this.email = email;
    }

    @Override
    public void onEvent(String eventType, Map<String, Object> data) {
        System.out.println("  [Email] Sending to " + email + ": File " +
                          data.get("filename") + " was " + eventType + "ed");
    }
}

// Stock market example

class Stock implements Subject {
    private String symbol;
    private double price;
    private List<Observer> investors = new ArrayList<>();

    public Stock(String symbol, double initialPrice) {
        this.symbol = symbol;
        this.price = initialPrice;
    }

    @Override
    public void attach(Observer observer) {
        investors.add(observer);
    }

    @Override
    public void detach(Observer observer) {
        investors.remove(observer);
    }

    @Override
    public void notifyObservers(String event, Object data) {
        for (Observer observer : investors) {
            observer.update(event, data);
        }
    }

    public void setPrice(double newPrice) {
        double oldPrice = this.price;
        this.price = newPrice;
        String direction = newPrice > oldPrice ? "UP" : "DOWN";
        double change = Math.abs(newPrice - oldPrice);

        System.out.println("  [" + symbol + "] Price changed: $" +
                          String.format("%.2f", oldPrice) + " -> $" +
                          String.format("%.2f", newPrice));

        Map<String, Object> data = new HashMap<>();
        data.put("symbol", symbol);
        data.put("oldPrice", oldPrice);
        data.put("newPrice", newPrice);
        data.put("change", change);
        data.put("direction", direction);

        notifyObservers("PRICE_CHANGE", data);
    }

    public String getSymbol() { return symbol; }
    public double getPrice() { return price; }
}

class Investor implements Observer {
    private String name;

    public Investor(String name) {
        this.name = name;
    }

    @Override
    @SuppressWarnings("unchecked")
    public void update(String event, Object data) {
        Map<String, Object> stockData = (Map<String, Object>) data;
        String symbol = (String) stockData.get("symbol");
        double newPrice = (Double) stockData.get("newPrice");
        String direction = (String) stockData.get("direction");

        System.out.println("  [Investor " + name + "] " + symbol + " is " +
                          direction + "! New price: $" + String.format("%.2f", newPrice));
    }
}

public class MainObserver {
    public static void main(String[] args) {
        System.out.println("=== Observer Pattern Demonstration ===\n");

        // News agency example
        System.out.println("--- 1. News Agency with Multiple Channels ---");

        NewsAgency agency = new NewsAgency();

        NewsChannel cnn = new NewsChannel("CNN");
        NewsWebsite bbc = new NewsWebsite("BBC.com");
        MobileApp newsApp = new MobileApp("NewsApp");

        agency.attach(cnn);
        agency.attach(bbc);
        agency.attach(newsApp);

        agency.publishNews("Breaking: Major tech announcement today!");
        System.out.println();

        agency.detach(cnn);
        agency.publishNews("Update: More details emerging...");
        System.out.println();

        // Event-based system
        System.out.println("--- 2. File Editor with Event System ---");

        FileEditor editor = new FileEditor();

        LoggingListener logger = new LoggingListener("activity.log");
        EmailListener emailer = new EmailListener("admin@example.com");

        editor.getEvents().subscribe("open", logger);
        editor.getEvents().subscribe("save", logger);
        editor.getEvents().subscribe("save", emailer);

        editor.openFile("document.txt");
        editor.saveFile();
        editor.closeFile();
        System.out.println();

        // Stock market example
        System.out.println("--- 3. Stock Market Observers ---");

        Stock apple = new Stock("AAPL", 150.00);
        Stock google = new Stock("GOOGL", 2800.00);

        Investor john = new Investor("John");
        Investor jane = new Investor("Jane");

        apple.attach(john);
        apple.attach(jane);
        google.attach(jane);

        System.out.println();
        apple.setPrice(155.50);
        System.out.println();
        google.setPrice(2750.00);
        System.out.println();
        apple.setPrice(152.30);

        System.out.println("\n=== Summary ===");
        System.out.println("Observer pattern benefits:");
        System.out.println("  - Loose coupling between subject and observers");
        System.out.println("  - Support for broadcast communication");
        System.out.println("  - Dynamic relationships at runtime");
        System.out.println("\nVariations:");
        System.out.println("  - Push model: subject sends data with notification");
        System.out.println("  - Pull model: observers query subject for data");
        System.out.println("  - Event-based: multiple event types");
    }
}

File: typescript/design_pattern/observer/main_observer.ts
/**
 * Observer Design Pattern in TypeScript
 *
 * The Observer pattern defines a one-to-many dependency between objects
 * so that when one object changes state, all its dependents are notified
 * and updated automatically.
 */

// Observer interface - defines the update method that all observers must implement
interface Observer {
    update(stockSymbol: string, price: number): void;
    getName(): string;
}

// Subject interface - defines methods for managing observers
interface Subject {
    attach(observer: Observer): void;
    detach(observer: Observer): void;
    notify(): void;
}

// Concrete Subject - Stock Market that notifies observers of price changes
class StockMarket implements Subject {
    private observers: Observer[] = [];
    private stocks: Map<string, number> = new Map();

    public attach(observer: Observer): void {
        const isExist = this.observers.includes(observer);
        if (isExist) {
            console.log(`[Line 28] StockMarket: ${observer.getName()} is already subscribed.`);
            return;
        }
        this.observers.push(observer);
        console.log(`[Line 32] StockMarket: ${observer.getName()} has been subscribed.`);
    }

    public detach(observer: Observer): void {
        const observerIndex = this.observers.indexOf(observer);
        if (observerIndex === -1) {
            console.log(`[Line 37] StockMarket: ${observer.getName()} was not found.`);
            return;
        }
        this.observers.splice(observerIndex, 1);
        console.log(`[Line 41] StockMarket: ${observer.getName()} has been unsubscribed.`);
    }

    public notify(): void {
        console.log(`[Line 45] StockMarket: Notifying ${this.observers.length} observer(s)...`);
        for (const observer of this.observers) {
            for (const [symbol, price] of this.stocks) {
                observer.update(symbol, price);
            }
        }
    }

    // Business logic method to update stock prices
    public setStockPrice(symbol: string, price: number): void {
        const oldPrice = this.stocks.get(symbol);
        this.stocks.set(symbol, price);

        if (oldPrice !== undefined) {
            const change = ((price - oldPrice) / oldPrice * 100).toFixed(2);
            console.log(`[Line 59] StockMarket: ${symbol} price updated from $${oldPrice.toFixed(2)} to $${price.toFixed(2)} (${change}%)`);
        } else {
            console.log(`[Line 61] StockMarket: ${symbol} added at $${price.toFixed(2)}`);
        }

        this.notify();
    }

    public getObserverCount(): number {
        return this.observers.length;
    }
}

// Concrete Observer 1 - Individual Investor
class IndividualInvestor implements Observer {
    private name: string;
    private portfolio: Map<string, number> = new Map();
    private alertThreshold: number;

    constructor(name: string, alertThreshold: number = 5) {
        this.name = name;
        this.alertThreshold = alertThreshold;
        console.log(`[Line 79] IndividualInvestor: ${name} created with ${alertThreshold}% alert threshold`);
    }

    public update(stockSymbol: string, price: number): void {
        const previousPrice = this.portfolio.get(stockSymbol);

        if (previousPrice !== undefined) {
            const changePercent = ((price - previousPrice) / previousPrice) * 100;
            if (Math.abs(changePercent) >= this.alertThreshold) {
                console.log(`[Line 87] ${this.name}: ALERT! ${stockSymbol} changed by ${changePercent.toFixed(2)}% (now $${price.toFixed(2)})`);
            } else {
                console.log(`[Line 89] ${this.name}: ${stockSymbol} updated to $${price.toFixed(2)} (within threshold)`);
            }
        } else {
            console.log(`[Line 92] ${this.name}: Tracking new stock ${stockSymbol} at $${price.toFixed(2)}`);
        }

        this.portfolio.set(stockSymbol, price);
    }

    public getName(): string {
        return this.name;
    }
}

// Concrete Observer 2 - Institutional Investor
class InstitutionalInvestor implements Observer {
    private name: string;
    private watchlist: string[];
    private priceHistory: Map<string, number[]> = new Map();

    constructor(name: string, watchlist: string[]) {
        this.name = name;
        this.watchlist = watchlist;
        console.log(`[Line 111] InstitutionalInvestor: ${name} created, watching: ${watchlist.join(", ")}`);
    }

    public update(stockSymbol: string, price: number): void {
        if (!this.watchlist.includes(stockSymbol)) {
            console.log(`[Line 116] ${this.name}: Ignoring ${stockSymbol} (not in watchlist)`);
            return;
        }

        const history = this.priceHistory.get(stockSymbol) || [];
        history.push(price);
        this.priceHistory.set(stockSymbol, history);

        const avgPrice = history.reduce((a, b) => a + b, 0) / history.length;
        console.log(`[Line 124] ${this.name}: ${stockSymbol} = $${price.toFixed(2)} | Avg: $${avgPrice.toFixed(2)} | Data points: ${history.length}`);
    }

    public getName(): string {
        return this.name;
    }
}

// Concrete Observer 3 - News Agency
class NewsAgency implements Observer {
    private name: string;
    private significantChangeThreshold: number;
    private lastReportedPrices: Map<string, number> = new Map();

    constructor(name: string, significantChangeThreshold: number = 10) {
        this.name = name;
        this.significantChangeThreshold = significantChangeThreshold;
        console.log(`[Line 140] NewsAgency: ${name} created with ${significantChangeThreshold}% significance threshold`);
    }

    public update(stockSymbol: string, price: number): void {
        const lastPrice = this.lastReportedPrices.get(stockSymbol);

        if (lastPrice !== undefined) {
            const changePercent = ((price - lastPrice) / lastPrice) * 100;
            if (Math.abs(changePercent) >= this.significantChangeThreshold) {
                const direction = changePercent > 0 ? "SURGES" : "PLUMMETS";
                console.log(`[Line 150] ${this.name}: BREAKING NEWS - ${stockSymbol} ${direction} ${Math.abs(changePercent).toFixed(1)}% to $${price.toFixed(2)}!`);
                this.lastReportedPrices.set(stockSymbol, price);
            } else {
                console.log(`[Line 153] ${this.name}: ${stockSymbol} at $${price.toFixed(2)} (no significant news)`);
            }
        } else {
            console.log(`[Line 156] ${this.name}: Initial report - ${stockSymbol} trading at $${price.toFixed(2)}`);
            this.lastReportedPrices.set(stockSymbol, price);
        }
    }

    public getName(): string {
        return this.name;
    }
}

// Demonstration
function main(): void {
    console.log("=== Observer Pattern Demonstration ===");
    console.log("=== Stock Market Monitoring System ===\n");

    // Create the subject (stock market)
    const stockMarket = new StockMarket();
    console.log("[Line 172] Created StockMarket instance\n");

    // Create observers
    console.log("--- Creating Observers ---");
    const alice = new IndividualInvestor("Alice", 5);
    const bob = new IndividualInvestor("Bob", 3);
    const hedgeFund = new InstitutionalInvestor("AlphaCapital", ["AAPL", "GOOGL"]);
    const newsAgency = new NewsAgency("MarketWatch", 8);
    console.log("");

    // Subscribe observers to the stock market
    console.log("--- Subscribing Observers ---");
    stockMarket.attach(alice);
    stockMarket.attach(bob);
    stockMarket.attach(hedgeFund);
    stockMarket.attach(newsAgency);
    console.log(`[Line 187] Total observers: ${stockMarket.getObserverCount()}\n`);

    // Try to subscribe an existing observer
    console.log("--- Attempting Duplicate Subscription ---");
    stockMarket.attach(alice);
    console.log("");

    // Initial stock prices
    console.log("--- Setting Initial Stock Prices ---");
    stockMarket.setStockPrice("AAPL", 150.00);
    console.log("");

    stockMarket.setStockPrice("GOOGL", 2800.00);
    console.log("");

    stockMarket.setStockPrice("TSLA", 250.00);
    console.log("");

    // Update stock prices - small change
    console.log("--- Small Price Update (within thresholds) ---");
    stockMarket.setStockPrice("AAPL", 152.00);
    console.log("");

    // Update stock prices - significant change
    console.log("--- Significant Price Update ---");
    stockMarket.setStockPrice("AAPL", 165.00);
    console.log("");

    // Unsubscribe Bob
    console.log("--- Unsubscribing Bob ---");
    stockMarket.detach(bob);
    console.log(`[Line 217] Remaining observers: ${stockMarket.getObserverCount()}\n`);

    // Another update after Bob unsubscribed
    console.log("--- Price Update After Bob Unsubscribed ---");
    stockMarket.setStockPrice("GOOGL", 3100.00);
    console.log("");

    // Try to unsubscribe non-existent observer
    console.log("--- Attempting to Unsubscribe Non-existent Observer ---");
    const phantom = new IndividualInvestor("Phantom", 1);
    stockMarket.detach(phantom);
    console.log("");

    // Unsubscribe all remaining observers
    console.log("--- Unsubscribing All Remaining Observers ---");
    stockMarket.detach(alice);
    stockMarket.detach(hedgeFund);
    stockMarket.detach(newsAgency);
    console.log(`[Line 235] Final observer count: ${stockMarket.getObserverCount()}\n`);

    // Final update with no observers
    console.log("--- Price Update With No Observers ---");
    stockMarket.setStockPrice("TSLA", 275.00);
    console.log("");

    console.log("=== End of Demonstration ===");
}

main();
